Support exchange spooling on GCS

linzebing opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?

New feature.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

trino-exchange-filesystem

> How would you describe this change to a non-technical end user or system administrator?

This PR adds support for exchange spooling on GCS. GCS is mostly S3-compatible, except for two minor incompatibilities.

An example `exchange-manager.properties`:
```
exchange-manager.name=filesystem
exchange.base-directories=gs://your-bucket-name
exchange.s3.region=us-west-1
exchange.s3.aws-access-key=your-google-access-key-id
exchange.s3.aws-secret-key=your-google-access-key-secret
exchange.s3.endpoint=https://storage.googleapis.com
```

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
(x) Documentation issue #issuenumber is filed, and can be handled later.
https://github.com/trinodb/trino/issues/12467

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Support exchange spooling on Google Cloud Storage.
* Dropped exchange spooling support for legacy S3 schemes s3n:// and s3a://.
```

electrum commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchangeModule.java:
```diff
@@ -53,7 +53,7 @@ protected void setup(Binder binder)
             binder.bind(FileSystemExchangeStorage.class).to(LocalFileSystemExchangeStorage.class).in(Scopes.SINGLETON);
         }
-        else if (ImmutableSet.of("s3", "s3a", "s3n").contains(scheme)) {
+        else if (ImmutableSet.of("s3", "s3a", "s3n", "gs").contains(scheme)) {
```
Unrelated to this PR, but we should limit S3 to just `s3` as the other schemes are legacy for Hive.

electrum commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -195,8 +195,15 @@ public ListenableFuture<Void> deleteRecursively(URI dir)
+                        // https://cloud.google.com/storage/docs/migrating#methods-comparison
+                        return deleteObjectsOneAtATime(bucketName, keys.build());
+                    }
+                    else {
```
`else` is redundant

electrum commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -319,17 +326,29 @@ private ListObjectsV2Publisher listObjectsRecursively(URI dir)
+        }).collect(toImmutableList()))));
+    }
+
+    private ListenableFuture<Void> deleteObjectsOneAtATime(String bucketName, List<String> keys)
```
We can use batched requests for S3: https://cloud.google.com/storage/docs/batch

If there are thousands of files, trying to delete them individually seems like a bad idea.

linzebing replied:
I see this one too, need to search for GCS APIs to do so

linzebing replied:
Technically they are not strict blockers, but I'm still weighing on them.

arhimondr replied:
With an average request latency of `~10ms` (should we expect anything lower than that on average?) it will take half a minute to delete a directory for a normal exchange (50 tasks * 50 partitions = 2500) objects. Also I'm not sure if GCP won't start throttling us heavily because of this.

> Google's API is synchronous, whereas we want to delete asynchronously

Hmm, they really don't provide a non blocking SDK? Technically it's just an HTTP based REST API. Maybe we can use a simple HTTP client (e.g.: airlift-http-client) to issue these requests?

linzebing replied:
sync/async API is fine, we can always use a background thread to do this.

What makes me disturbed is that with the introduction of GCS client, we will need to separate credentials code. But seems that's the best thing I can do right now.

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -107,7 +107,7 @@
         implements FileSystemExchangeStorage
 {
-    private static final String DIRECTORY_SUFFIX = "_$folder$";
+    private static final String GCS_SCHEME = "gs";
```
I would rather determine operation mode when we setup `S3FileSystemExchangeStorage` instead of looking at schema of directory here and there.

You can do that by defining enum which denotes operation mode and pass the value to constructor:
```java

    public enum CompatibilityMode {
        AWS,
        GCP
    }

    @Inject
    public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, ExchangeS3Config config, CompatibilityMode compatibilityMode)
```

And then in `FileSystemExchangeModule` you can do:
```java
            S3FileSystemExchangeStorage.CompatibilityMode compatibilityMode = scheme.equals("gs") ? GCP : AWS;
            binder.bind(S3FileSystemExchangeStorage.CompatibilityMode.class).toInstance(compatibilityMode);
```

mosabua commented:
This will need docs .. please work with @colebow on adding this.

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -144,6 +163,21 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
                 config.getAsyncClientMaxPendingConnectionAcquires(),
                 config.getConnectionAcquisitionTimeout());
+
+        if (compatibilityMode.equals(CompatibilityMode.GCP)) {
```
nit: static import GCP

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -117,8 +131,13 @@
 
+    // GCS specific
+    private final Optional<Storage> gcsClient;
+    private final ListeningExecutorService gcsDeleteExecutor;
```
nit: mark as `@Nullable`

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -190,15 +224,26 @@ public ListenableFuture<Void> createEmptyFile(URI file)
-                    return deleteObjects(getBucketName(dir), keys.build());
-                },
-                directExecutor())));
+        if (gcsClient.isPresent()) {
```
nit: I would assign `compatibilityMode` to field and use it for dispatching (and only `verify` that `gcsClient.isPresent`. But I do not feel super strongly about that as so far `delete` impl is only GCS specific part.

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -190,15 +224,26 @@ public ListenableFuture<Void> createEmptyFile(URI file)
+                StorageBatch batch = storage.batch();
+                Page<Blob> blobs = storage.list(getBucketName(dir), Storage.BlobListOption.prefix(keyFromUri(dir)));
+                for (Blob blob : blobs.iterateAll()) {
+                    batch.delete(blob.getBlobId());
```
Are there any limits on batch size? Should we create multiple batches if `blobs` is to big?

linzebing replied:
GCP Java JDK already handled for us https://github.com/googleapis/java-storage/blob/2fd3378eaffb34ed0bb1e6f377d81bb6ccecf030/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L197-L210

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -574,17 +619,17 @@ private void fillBuffer()
                             .key(key)
                             .bucket(bucketName)
-                            .partNumber(partNumber);
+                            .range("bytes=" + fileOffset + "-" + (fileOffset + length - 1));
```
Is that still OK for AWS (e.g. doesn't it impose some performance degradation)?

linzebing replied:
<img width="1166" alt="image" src="https://user-images.githubusercontent.com/9404831/169119789-2112466a-1863-44c5-b7dd-776c43b71104.png">
To my understanding, they are equivalent. Previously we decided to use `partNumber` was to avoid accidentally reading across part boundaries. However, given that GCP is not compatible with partNumber, and community also sees incompatibilities (e.g. Netapp Storage Grid), I guess we may simply use byte range instead.

linzebing commented:
CI: https://github.com/trinodb/trino/issues/12413

arhimondr commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -116,16 +133,24 @@
+    // GCS specific
+    private final Optional<Storage> gcsClient;
+    @Nullable
+    private final ListeningExecutorService gcsDeleteExecutor;
```
nit: Maybe better define it as optional?

arhimondr commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -144,6 +169,21 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
+            else {
+                this.gcsClient = Optional.of(StorageOptions.getDefaultInstance().getService());
+            }
+            this.gcsDeleteExecutor = MoreExecutors.listeningDecorator(newCachedThreadPool(threadsNamed("gcs-delete-%s")));
```
This can in theory grow indefinitely. In `FileSystemExchange#close` we call `deleteRecursively` in a loop for each task. This may result in huge spikes in number of threads (hundreds or even thousands). I would recommend going with a bounded executor with the number of threads set to desired concurrency, e.g.:

```
        ThreadPoolExecutor executor = new ThreadPoolExecutor(
                maximumConcurrency,
                maximumConcurrency,
                10, SECONDS,
                new LinkedBlockingQueue<>(),
                threadsNamed("gcs-delete-%s"));
```

I don't know what value do we want to pick for max concurrency though, maybe 50? 100?

Also it is a generally good idea to set `executor.allowCoreThreadTimeOut(true);` to let the inactive threads be reclaimed after a spike. 

It might also be reasonable to improve `FileSystemExchange#close` to batch delete requests across multiple partitions.

arhimondr commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -190,15 +230,27 @@ public ListenableFuture<Void> createEmptyFile(URI file)
-                directExecutor())));
+        if (compatibilityMode.equals(GCP)) {
+            // GCS is not compatible with S3's multi-object delete API https://cloud.google.com/storage/docs/migrating#methods-comparison
+            verify(gcsClient.isPresent() && gcsDeleteExecutor != null);
```
nit: with `optional` it's usually more elegant. You can do something like

```
gcsClient.orElseThrow(()-> new IllegalStateException("gcsClient is expected to be initialized"))
```

linzebing commented:
I have changed to batch the batch deletes, such to minimize API calls to GCS.

Wonder if we should do the same for Azure and S3. Currently we are deleting a task output directory at a time. In theory, we can do something similar to GCS, collect all the objects into a list, and batch delete them. @arhimondr @losipiuk

losipiuk commented:
> I have changed to batch the batch deletes, such to minimize API calls to GCS.
> 
> Wonder if we should do the same for Azure and S3. Currently we are deleting a task output directory at a time. In theory, we can do something similar to GCS, collect all the objects into a list, and batch delete them. @arhimondr @losipiuk

Would make sense IMO. Good catch.

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -144,6 +167,21 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
                 config.getAsyncClientMaxPendingConnectionAcquires(),
                 config.getConnectionAcquisitionTimeout());
+
+        if (compatibilityMode.equals(GCP)) {
```
nit: you can use `==` for enums

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -190,15 +228,27 @@ public ListenableFuture<Void> createEmptyFile(URI file)
-                    return deleteObjects(getBucketName(dir), keys.build());
-                },
-                directExecutor())));
+        if (compatibilityMode.equals(GCP)) {
```
ditto

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchangeModule.java:
```diff
@@ -53,11 +55,13 @@ protected void setup(Binder binder)
         if (scheme == null || scheme.equals("file")) {
             binder.bind(FileSystemExchangeStorage.class).to(LocalFileSystemExchangeStorage.class).in(Scopes.SINGLETON);
         }
-        else if (ImmutableSet.of("s3", "s3a", "s3n").contains(scheme)) {
```
add entry to the release notes. Is there some doc to be updated?

linzebing replied:
I have updated https://github.com/trinodb/trino/pull/12472 accordingly

arhimondr reviewed (approved):
LGTM % comment

losipiuk commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -144,6 +167,21 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
+            else {
+                this.gcsClient = Optional.of(StorageOptions.getDefaultInstance().getService());
+            }
+            this.gcsDeleteExecutor = Optional.of(MoreExecutors.listeningDecorator(newCachedThreadPool(threadsNamed("gcs-delete-%s"))));
```
I agree with small comment that we use `cachedThreadPool` very commonly through the codebase. So probably one less/one more does not make much change :)

arhimondr replied:
Yeah, I've also noticed that. Maybe I'm being paranoid, but somehow it makes me feel anxious when I think of a thread pool that can grow indefinitely :-)

linzebing commented:
Addressed comments. On batching deletes for S3 and Azure, decided to do it in a separate PR. It's a bit more complex than I thought as I need to deal with multiple buckets

arhimondr commented on plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -144,6 +169,24 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
+            else {
+                this.gcsClient = Optional.of(StorageOptions.getDefaultInstance().getService());
+            }
+            this.gcsDeleteExecutor = Optional.of(MoreExecutors.listeningDecorator(new ThreadPoolExecutor(0, 100,
```
nit: one parameter per line, static import `listeningDecorator`

arhimondr replied:
Also set the core pool size to 100, otherwise it will keep running only a single thread until the queue is full

linzebing replied:
Ah you are right. I'm using a `SynchronousQueue` here, which basically has a size of 0, and if concurrent tasks exceed 100, rejection will happen. Your suggestion above is better.
